File: db/src/main/java/com/ouattararomuald/saviezvousque/db/DbModule.java
package com.ouattararomuald.saviezvousque.db;

import android.content.Context;
import com.ouattararomuald.saviezvousque.db.adapters.LocalDateTimeAdapter;
import com.ouattararomuald.saviezvousque.db.daos.CategoryDao;
import com.ouattararomuald.saviezvousque.db.daos.PostDao;
import com.squareup.sqldelight.android.AndroidSqliteDriver;
import com.squareup.sqldelight.db.SqlDriver;
import dagger.Module;
import dagger.Provides;

@Module
abstract class DbModule {

  @Provides
  static PostQueries postQueries(Database database) {
    return database.getPostQueries();
  }

  @Provides
  static CategoryQueries categoryQueries(Database database) {
    return database.getCategoryQueries();
  }

  @Provides
  static Database sqlDelightDatabase(SqlDriver driver, LocalDateTimeAdapter localDateTimeAdapter) {
    return Database.Companion.invoke(driver,
        new Post.Adapter(localDateTimeAdapter, localDateTimeAdapter));
  }

  @Provides
  static LocalDateTimeAdapter localDateTimeAdapter() {
    return new LocalDateTimeAdapter();
  }

  @Provides
  static SqlDriver sqlDriver(Context context, String databaseName) {
    return new AndroidSqliteDriver(Database.Companion.getSchema(), context, databaseName);
  }

  @Provides
  static CategoryDao categoryDao(CategoryQueries categoryQueries) {
    return new CategoryDao(categoryQueries);
  }

  @Provides
  static PostDao postDao(PostQueries postQueries) {
    return new PostDao(postQueries);
  }

  @Provides
  public static FeedRepository feedRepository(
      CategoryDao categoryDao,
      PostDao postDao
  ) {
    return new FeedRepository(categoryDao, postDao);
  }
}

File: db/src/main/java/com/ouattararomuald/saviezvousque/db/DbModule.kt
package com.ouattararomuald.saviezvousque.db

import android.content.Context
import com.ouattararomuald.saviezvousque.db.Database.Companion.Schema
import com.ouattararomuald.saviezvousque.db.adapters.LocalDateTimeAdapter
import com.squareup.sqldelight.android.AndroidSqliteDriver
import com.squareup.sqldelight.db.SqlDriver
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@InstallIn(SingletonComponent::class)
@Module
object DbModule {
  @Singleton
  @Provides
  fun localDateTimeAdapter(): LocalDateTimeAdapter = LocalDateTimeAdapter()

  @Singleton
  @Provides
  fun sqlDriver(@ApplicationContext context: Context): SqlDriver = AndroidSqliteDriver(
    Schema,
    context,
    BuildConfig.DATABASE_NAME
  )

  @Singleton
  @Provides
  fun sqlDelightDatabase(
    driver: SqlDriver,
    localDateTimeAdapter: LocalDateTimeAdapter
  ): Database = Database(
    driver,
    Post.Adapter(
      localDateTimeAdapter,
      localDateTimeAdapter
    )
  )

  @Provides fun postQueries(database: Database): PostQueries = database.postQueries

  @Provides fun categoryQueries(database: Database): CategoryQueries = database.categoryQueries
}